src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs

```diff
@@ -432,30 +442,43 @@ public DfaMatchingState<TElement> MkState(SymbolicRegexNode<TElement> node, uint
                 }
                 else
                 {
-                    state = MakeNewState(s);
+                    state = MakeNewState(s, capturing);
                 }
             }
 
             return state;
         }
 
-        private DfaMatchingState<TElement> MakeNewState(DfaMatchingState<TElement> state)
+        private DfaMatchingState<TElement> MakeNewState(DfaMatchingState<TElement> state, bool capturing)
         {
             lock (this)
             {
-                state.Id = _stateCache.Count;
-                _stateCache.Add(state);
+                HashSet<DfaMatchingState<TElement>> cache = capturing ? _stateCache : _capturingStateCache;
+                state.Id = cache.Count;
+                cache.Add(state);
 
-                Debug.Assert(_statearray is not null);
+                Debug.Assert(_statearray is not null && _capturingStatearray is not null);
 
-                if (state.Id == _statearray.Length)
+                if (capturing)
                 {
-                    int newsize = _statearray.Length + 1024;
-                    Array.Resize(ref _statearray, newsize);
-                    Array.Resize(ref _delta, newsize << _mintermsCount);
-                    Array.Resize(ref _capturingDelta, newsize << _mintermsCount);
+                    if (state.Id == _capturingStatearray.Length)
+                    {
+                        int newsize = _capturingStatearray.Length + 1024;
+                        Array.Resize(ref _capturingStatearray, newsize);
+                        Array.Resize(ref _capturingDelta, newsize << _mintermsCount);
+                    }
+                    _capturingStatearray[state.Id] = state;
+                }
+                else
+                {
+                    if (state.Id == _statearray.Length)
+                    {
+                        int newsize = _statearray.Length + 1024;
+                        Array.Resize(ref _statearray, newsize);
+                        Array.Resize(ref _delta, newsize << _mintermsCount);
```

Comment:
I was going to suggest we use ref locals to avoid the duplication between the if/else branches, but the fact that _delta and _capturingDelta are of different types makes that more challenging.  I guess the duplication is probably better than alternatives.

Reply:
Yes, I tried to write it with ref locals before I realized the types don't match.